projects/vdk-plugins/vdk-data-source-git/src/vdk/plugin/data_source_git/git_source.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2021-2023 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+import os
+import tempfile
+from typing import Iterator
+from typing import List
+from typing import Optional
+
+from dulwich import porcelain
+from vdk.plugin.data_source_git.utils import detect_language
+from vdk.plugin.data_source_git.utils import is_test_file
+from vdk.plugin.data_sources.config import config_class
+from vdk.plugin.data_sources.config import config_field
+from vdk.plugin.data_sources.data_source import DataSourcePayload
+from vdk.plugin.data_sources.data_source import IDataSource
+from vdk.plugin.data_sources.data_source import (
+    IDataSourceConfiguration,
+)
+from vdk.plugin.data_sources.data_source import IDataSourceStream
+from vdk.plugin.data_sources.factory import data_source
+from vdk.plugin.data_sources.state import IDataSourceState
+
+DESCRIPTION = """Git data source.
+Extract content from Git repositories and associated file metadata.
+"""
+
+
+@config_class(name="git", description=DESCRIPTION)
+class GitDataSourceConfiguration(IDataSourceConfiguration):
+    git_url: str = config_field(description="Git URL that would be cloned. ")
+    git_ssh_key: Optional[str] = config_field(
+        description="SSH key to use when cloning the repo."
+        "Leave empty if no authentication is needed",
+        default="",
+    )
+
+
+@data_source(name="git", config_class=GitDataSourceConfiguration)
+class GitDataSource(IDataSource):
```

Comment:
This datasource will always only have a single stream?

Reply:
At this iteration of the data source implementation. All singer data sources are really relational data sources (json / dictionary) I needed a data source that is blobs of data so that's why I developed that. To find out the limitations of our ingestion . Eventually it might make sense to allow users to configure streams - maybe branches, or directories or something else. But not at this first iteration. 

It's development status is pre-alpha currently or maybe alphase as per https://martin-thoma.com/software-development-stages/


PS : The main limitation I found is that we want the payloads to be json serializable so we don't really accept "bytes" in the payload.